### PR TITLE
Fix receiver replication accounting on parents

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary
- Centralize receiver replication claim/release ownership, including chart teardown.
- Prevent counter underflow and stale replication contributions.
- Report inbound replication completion from the connection’s chart cohort, keeping status, instance count, and completion consistent.
- Add deterministic unit coverage for replication-accounting races and lifecycle paths.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved receiver replication status accuracy using consistent accounting snapshots.
  - Prevented duplicate chart events from changing replication progress or completion metrics.
  - Prevented progress from incorrectly reaching 100% while charts remain outstanding.
  - Ensured replication progress is released when charts are removed, become obsolete, or replication resets.
  - Prevented invalid releases from causing counter underflows or corrupting status.
  - Improved replication handling during concurrent chart updates, disconnects, and stream transitions.

### CODE REVIEW

File: src/daemon/unit_test.c [hunk 1]
```diff
@@ -2069,6 +2069,440 @@ static int test_rrdset_rejects_invalid_update_every(void) {
...
+    RRDSET *st;
+    size_t iterations;
+    size_t claims;
+};
+
+static void receiver_replication_claim_thread(void *ptr) {
+    struct receiver_replication_race_thread *t = ptr;
+
+    for(size_t i = 0; i < t->iterations; i++) {
+        if(rrdhost_receiver_replication_claim(t->st))
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/daemon/unit_test.c:2435]: Valid, fixed. The stress pass runs 8x500 claim/release pairs and `charts_started` is append-only by design, so every won claim grew it while the balanced `charts` counter returned to baseline. `test_receiver_replication_ownership_under_concurrency()` now saves `charts_started` at entry and restores it before returning, covering both the deterministic part 1 and the stress part 2 - the same convention the teardown and completion tests in this file already follow.

File: src/database/rrdhost-status.c [hunk 2]
```diff
@@ -176,13 +181,23 @@ static inline RRDHOST_INGEST_STATUS rrdhost_status_ingest(RRDHOST *host, RRDHOST
...
+            // on the API. This is the whole reason the helper needs no re-read.
+            rrdhost_receiver_lock(host);
+            replication_completion = rrdhost_receiver_replication_completion(host, &replicating_instances);
+            rrdhost_receiver_unlock(host);
 
-        else
-            status = RRDHOST_INGEST_STATUS_ONLINE;
+            collected_metrics = __atomic_load_n(&host->collected.metrics_count, __ATOMIC_RELAXED);
+
+            if(replicating_instances > 0 || !collected_metrics)
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/database/rrdhost-status.c:194]: Correct on the facts, and the comment was the thing that was wrong - not the code.

The `!collected_metrics` clause is pre-existing and deliberate: a connected child stays in `replicating` until its first metric arrives. In that window nothing is outstanding, so the completion helper returns its empty-cohort 100.0, and `instances` is 0. So yes, the triple is reachable, and the new comment claimed the single snapshot prevented it. It does not, and it was never meant to - what the snapshot buys is that the status decision and the reported instances/completion cannot disagree with each other (a decision taken on a non-zero instance count paired with a later, already-drained 100).

Reworded the comment at rrdhost-status.c to say exactly that, and to name the pre-first-metric window as the intended state rather than a torn read. Left the behaviour alone: `remaining == 0 -> 100.0` is the helper's pinned contract (`test_receiver_replication_completion`, case 'idle host, nothing ever claimed'), and making an empty cohort read 0 instead would make a child that has been online for hours and never replicated report `InReplCompletion: 0`, which is worse than the few seconds this affects.

File: src/database/rrdset.h [hunk 3]
```diff
@@ -83,6 +83,36 @@ typedef enum __attribute__ ((__packed__)) rrdset_flags {
...
 #define rrdset_is_discoverable(st) (rrdset_is_replicating(st) || !rrdset_flag_check(st, RRDSET_FLAG_OBSOLETE))
 
+// --------------------------------------------------------------------------------------------------------------------
+// Receiver replication ownership.
+//
+// The SINGLE implementation of the receiver-replication claim and release. Every lifecycle site uses
+// these - the parser's CHART_DEFINITION_END, both REPLAY_END branches, the connect/disconnect reset,
+// and chart teardown. Do NOT open-code the flag CAS plus the counter movement anywhere else: the
+// ownership invariant is only enforceable while it lives in one place. Contract and its two exceptions:
+// .agents/sow/specs/streaming-receiver-replication-accounting.md
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/database/rrdset.h]: Valid, fixed. That path is a gitignored, local-only working file - a shipped header must not send a maintainer somewhere they cannot open. The two legal decrement shapes are now stated inline in the header: either a release that cleared IN_PROGRESS in its own CAS old value, or the claim's rollback of the speculative increment the same thread just made on a duplicate CHART_DEFINITION_END (which does not clear the flag, because the chart still legitimately holds its contribution).

File: src/daemon/unit_test.c [hunk 4]
```diff
@@ -2069,6 +2069,440 @@ static int test_rrdset_rejects_invalid_update_every(void) {
...
+static int test_receiver_replication_counter_does_not_wrap(void) {
+    fprintf(stderr, "%s() running...\n", __FUNCTION__);
+
+    int rc = 0;
+    RRDHOST *host = localhost;
+
+    uint32_t saved = rrdhost_receiver_replicating_charts(host);
+    rrdhost_receiver_replicating_charts_zero(host);
+
+    uint32_t after = rrdhost_receiver_replicating_charts_minus_one(host);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/daemon/unit_test.c:2154]: Fair - documented rather than silenced. The log line is the whole point of the guard, so gating it on a unit-test build would delete the diagnostic value in exactly the configuration CI runs. Instead the test now prints, immediately before the deliberate release, that it expects one `STREAM REPLAY ERROR ... refused` line and that the line is the guard working. That puts the disclaimer adjacent to the error in the CI log rather than in a source file nobody is reading at that moment, and there is a matching source comment.

File: src/daemon/unit_test.c [hunk 5]
```diff
@@ -2069,6 +2069,440 @@ static int test_rrdset_rejects_invalid_update_every(void) {
...
+                "netdata", id, id, "netdata", NULL, "Unit Testing", "x", "unittest", NULL, 1,
+                nd_profile.update_every, RRDSET_TYPE_LINE);
+        }
+
+        uint32_t before = rrdhost_receiver_replicating_charts(host);
+#ifdef NETDATA_INTERNAL_CHECKS
+        size_t refusals_before = __atomic_load_n(&rrdhost_receiver_replication_refusals, __ATOMIC_RELAXED);
+#endif
+
+        struct receiver_replication_race_thread t[8];
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/daemon/unit_test.c]: Valid, fixed. `threads_count` and `charts_count` are now an `enum` and the arrays are declared `charts[charts_count]` / `t[threads_count]`, so the size and the loop bound cannot diverge. `enum` rather than `const size_t` on purpose: a `const size_t` is not a constant expression in C, so those declarations would have become VLAs. The values and the comment explaining the CI/ASAN sizing are unchanged, and the `%zu` that printed `threads_count` now casts, since the enum constant is an `int`.

**Comment by stelfrag** [comment]
@cubic-dev-ai please review again

**Comment by stelfrag** [comment]
@coderabbitai review

File: src/database/rrdhost.c [hunk 6]
```diff
@@ -946,6 +945,185 @@ RRDHOST *rrdhost_find_or_create(
...
+//
+// Ordering, which the whole accounting rests on: the claim-attempt cohort is incremented first, then
+// the live counter, and only then is RRDSET_FLAG_RECEIVER_REPLICATION_IN_PROGRESS published. A reader
+// that observes the published flag therefore always observes a counter already incremented to match.
+// Publishing first would make the flag visible while the contribution does not yet exist, and any
+// release site would then decrement something nobody owns.
+bool rrdhost_receiver_replication_claim(RRDSET *st) {
+    RRDHOST *host = st->rrdhost;
+
+    rrdhost_receiver_replicating_charts_started_plus_one(host);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/database/rrdhost.c]: Confirmed, and worse than the code claimed. Completion is (started - remaining) / started, so leaving `charts_started` raised while withdrawing `charts` lifts the numerator AND the denominator - the ratio goes UP. One duplicate on a single outstanding chart read 50%, N duplicates read N/(N+1). The 'inflates the denominator, so it under-reports' reasoning in rrdhost.c, rrdhost.h and the test case named 'duplicate claim attempt inflated it' was inverted, and a child re-sends CHART_DEFINITION_END on every dimension it adds to an already-replicating chart, so this was routine rather than a corner case. It reintroduced the exact defect the PR exists to remove.

Fixed by removing the reason the rollback was one-sided rather than by adding a lock. The two counters are now ONE 64-bit atomic word, `charts_started_and_remaining` - high half the cohort, low half the outstanding charts. The claim adds ((1<<32)|1) in a single atomic op and the duplicate rollback subtracts the same unit in a single atomic op, so a duplicate moves neither half and cannot move the figure. Because the pair is one word it can never be read torn, which also let the acquire/release pairing between the two counters go, and with it the `rrdhost_receiver_lock()` that rrdhost_status_ingest() was taking purely to keep the two-store reset from interleaving between the two loads. The reset is now a single store, and the `started < remaining` NAN branch is unreachable by construction rather than by argument.

Both decrement shapes - the release and the claim's withdrawal - share one guarded helper that refuses when the outstanding half is zero: on a packed word an unguarded subtract does not merely wrap the low half, it borrows out of the cohort and destroys the denominator. Added a regression test that claims a chart, re-claims the same chart, and asserts both halves, the completion figure and the instance count are unchanged. Comments, the removed test case, and the operator-facing note that repeated the inverted claim are all corrected.

[the remark this answers is not in the document]
Comment (stelfrag) [reply src/database/rrdhost.c]: Confirmed - same finding as the coderabbit thread on this line, see the detail there. Short version: leaving the cohort raised does not bias the figure downwards, it biases it UP, because completion is (started - remaining) / started and the duplicate lifts both terms.

Fixed by making the two counters a single 64-bit atomic word rather than by locking them together: the claim adds ((1<<32)|1) in one atomic op and the duplicate rollback subtracts the same unit in one atomic op, so both halves move together and a duplicate CHART_DEFINITION_END changes neither. That is also 'update the cohort atomically' as you suggested, without a lock on the claim path - and it removes the acquire/release pairing and the receiver lock in rrdhost_status_ingest() that only existed to read the two counters consistently.

File: src/daemon/unit_test.c [hunk 7]
```diff
@@ -2069,6 +2069,451 @@ static int test_rrdset_rejects_invalid_update_every(void) {
...
+        size_t claims = 0;
+        for(size_t i = 0; i < threads_count; i++) {
+            nd_thread_join(t[i].thread);
+            claims += t[i].claims;
+        }
+
+        // Every thread's last act is a release, so every chart is released by whoever touched it last;
+        // this only covers the case where a chart is left claimed.
+        for(size_t c = 0; c < charts_count; c++)
+            rrdhost_receiver_replication_release(charts[c], 0);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/daemon/unit_test.c]: Valid, fixed. Both assertions - the counter comparison and the refusal count - now run BEFORE that loop, and the loop is explicitly labelled post-assertion cleanup. You are right that it could erase the failure: a leaked contribution whose flag is still set would have been quietly released by the cleanup and the counter would have landed back on its baseline.

File: src/daemon/unit_test.c [hunk 8]
```diff
@@ -2069,6 +2069,451 @@ static int test_rrdset_rejects_invalid_update_every(void) {
...
+};
+
+static void receiver_replication_claim_thread(void *ptr) {
+    struct receiver_replication_race_thread *t = ptr;
+
+    for(size_t i = 0; i < t->iterations; i++) {
+        if(rrdhost_receiver_replication_claim(t->st))
+            t->claims++;
+
+        rrdhost_receiver_replication_release(t->st, 0);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/daemon/unit_test.c:2445]: Not taking this one - it would weaken the test, and the premise does not hold.

A release cannot steal anything. rrdhost_receiver_replication_release() decrements only when its own CAS old-value contained RRDSET_FLAG_RECEIVER_REPLICATION_IN_PROGRESS, i.e. only when THIS call is the one that cleared the flag. A call that does not win the clear is a no-op and touches no counter, so a worker whose claim returned false and that then calls release either wins the clear - in which case it legitimately owns that transition, exactly like the parser's REPLAY_END does for a chart it did not claim - or does nothing at all. The counter stays balanced either way, which is what the `after != before` assertion checks.

Releasing unconditionally is the point of the stress pass: it puts non-owning threads into the release path concurrently, which is the only thing that makes the refusal-count assertion meaningful. Move the release inside the `if` and every release becomes an owned one, refusals are trivially zero, and the pass stops exercising ownership under contention at all. The deterministic part 1 above it already pins the specific interleaving the ordering exists for; part 2's job is the interleavings a hand-written case did not think of.

Added a comment at that line saying so, so it does not get re-raised.

File: src/database/rrdhost.c [hunk 9]
```diff
@@ -946,6 +945,185 @@ RRDHOST *rrdhost_find_or_create(
...
+    rrdhost_receiver_replicating_charts_plus_one(host);
+
+    rrdhost_receiver_replication_race_pause(st);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/database/rrdhost.c]: False positive, on both halves - and the first half is not something this PR changes.

The chart cannot be freed in that window. `rrdset_free()` is reached only from `svc_rrdhost_cleanup_charts_marked_obsolete()` (daemon/service.c), behind `svc_rrdset_lock_for_deletion()`, which requires ALL of `rrdset_last_accessed_time_s(st)`, `st->last_updated.tv_sec` and `st->last_collected_time.tv_sec` to be older than `rrdset_free_obsolete_time_s` (default 3600s, netdata-conf-db.c:15), plus RRDSET_FLAG_OBSOLETE, plus a successful `spinlock_trylock(&st->destroy_lock)`, plus `dictionary_acquired_item_references(st_dfe.item) == 1`. A chart the parser is currently inside CHART_DEFINITION_END for was touched microseconds earlier by the CHART that set the scope, so the first condition alone is false by an hour. And `rrdset_free()` -> `dict_item_del()` only flags a still-referenced item and defers the callback, so even a racing holder does not free under us.

Second, the `st` dereference at the flag CAS is not introduced here. Before this PR the same function did `rrdset_flag_set_and_clear(st, ...)` and `rrdhost_receiver_replicating_charts_plus_one(st->rrdhost)` at the same point of the same command - see the diff of pluginsd_replication.c. What changed is the ORDER (increment then publish, instead of publish then increment), not whether `st` is touched. If the pointer could be freed under the parser, every `st` use in every parser command would be affected, not this line; `parser->user.st` is a bare pointer by design across the whole parser.

The orphaned-increment half depends on the free actually happening, so it falls with the first point. For what it is worth the interleaving itself - a release arriving after the increment but before the publish - is deliberately pinned by a deterministic test using a pause hook (test_receiver_replication_ownership_under_concurrency, part 1): the release correctly finds no IN_PROGRESS and consumes nothing, and the counter balances. That is exactly the ordering this arrangement exists to get right; publishing first is what would let a release decrement something nobody owns.

File: src/daemon/unit_test.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/daemon/unit_test.c:2445]: Same finding as an earlier thread on this line, and the answer is the same: not taking it, because the premise about stealing does not hold and the change would weaken the test.

A losing worker cannot steal anything. The decrement happens only inside rrdhost_receiver_replication_release(), and only when THAT call's own CAS old-value contained RRDSET_FLAG_RECEIVER_REPLICATION_IN_PROGRESS - i.e. only when it is the call that cleared the flag. So a losing claimer's release either wins the clear, in which case it legitimately owns that transition and consumes the contribution the winner published (an ownership transfer, the same shape as REPLAY_END releasing a chart it never claimed), or it does not win and touches no counter at all. Either way the counter stays balanced, which is what the `after != before` assertion checks, and the ownership API is doing exactly what it is meant to.

Releasing on every iteration is the point of the stress pass: it is what puts non-owning threads into the release path concurrently, and that is the only thing that makes the refusal-count assertion meaningful. Move the release inside the `if` and every release becomes an owned one, refusals go to trivially zero, and the pass stops exercising ownership under contention. The deterministic part 1 above it already pins the specific interleaving; part 2's job is the ones a hand-written case did not think of.

The source comment at that line has been expanded to say this, so it does not get raised a third time.

**Comment by stelfrag** [comment]
@cubic-dev-ai please review again

**Comment by stelfrag** [comment]
@coderabbitai review

**Comment by stelfrag** [comment]
@cubic-dev-ai review PR
